Build and install cookbooks and benchmarks

gassmoeller opened the pull request:
This PR allows to build all benchmark and cookbook plugins and install them alongside the rest of the benchmark and cookbook files. This should be useful for VMs and docker images as well as testing (when I enabled this option I immediately got some CMake warnings and errors that are fixed below (e.g. using the same project name in a benchmark and cookbook project). This will also allow us to enable building cookbooks and benchmark plugins on the testers, checking for warnings or compile problems.

I wouldnt like to compile the shared libraries always by default (they take some time and are often not needed), but on the other hand installing the examples without the built shared libraries feels a bit weird. This is why I decided to put them behind the same cmake option `ASPECT_INSTALL_EXAMPLES`, which is off by default for now.  Opinions?

Note that the change of the install location in `cmake/AspectConfig.cmake.in` works no matter if the plugin is compiled from the actual plugin directory (in which case the install path will evaluate to `$CMAKE_INSTALL_PREFIX/.`) or being included from the main build folder (because then `FILE(RELATIVE_PATH _relative_path ${CMAKE_SOURCE_DIR} ${CMAKE_CURRENT_LIST_DIR})` will evaluate to the relative path between main CMakeLists.txt file and the current plugin directory, e.g. `cookbooks/finite_strain`). It even works for all plugins outside of the main ASPECT, because for them `${CMAKE_SOURCE_DIR}` and `${CMAKE_CURRENT_LIST_DIR}` is simply the same, so that `make install` will try to install to `$CMAKE_INSTALL_PREFIX` as expected.

tjhei commented:
> This will also allow us to enable building cookbooks and benchmark plugins on the testers,

We already do this on Jenkins:
https://github.com/geodynamics/aspect/blob/56a1bcd14051c261d4d8a15426b6937c8053bbd8/Jenkinsfile#L113-L126

But we could simplify that mechanism with your PR. I am happy to do that (the makefile is somewhat messy).

tjhei commented on CMakeLists.txt:
```diff
@@ -793,14 +793,21 @@ INSTALL(DIRECTORY include/ ${CMAKE_BINARY_DIR}/include/
-  DESTINATION benchmarks
-  COMPONENT examples
-  FILES_MATCHING PATTERN "*")
+SET(ASPECT_INSTALL_EXAMPLES OFF CACHE BOOL "")
```
Can you add a description as the last argument here?

tjhei commented on benchmarks/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+# Step 2: Add all subdirectories of plugins.
+FOREACH(_file ${_cmakelists_benchmarks})
+  GET_FILENAME_COMPONENT(_dir ${_file} DIRECTORY)
+  SET(Aspect_DIR ${CMAKE_BINARY_DIR})
```
this statement can be moved outside the loop

tjhei commented on benchmarks/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+
+# Step 1: Find all of the benchmarks that have separate plugins.
+FILE(GLOB_RECURSE _cmakelists_benchmarks
+     ${CMAKE_SOURCE_DIR}/benchmarks/**/CMakeLists.txt)
```
```suggestion
     ${CMAKE_CURRENT_SOURCE_DIR}/**/CMakeLists.txt)
```
Does this work?

tjhei commented on cookbooks/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+# Step 2: Add all subdirectories of plugins.
+FOREACH(_file ${_cmakelists_cookbooks})
+  GET_FILENAME_COMPONENT(_dir ${_file} DIRECTORY)
+  SET(Aspect_DIR ${CMAKE_BINARY_DIR})
```
same here

tjhei commented on cookbooks/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+     ${CMAKE_SOURCE_DIR}/cookbooks/**/CMakeLists.txt)
+
+# Step 2: Add all subdirectories of plugins.
+FOREACH(_file ${_cmakelists_cookbooks})
```
I would use generic names (the 2 CMakeLists.txt files can basically be identical)

gassmoeller commented:
> Does this work?

Yes, thanks that simplifies it further. I addressed all comments, and the CMakeLists.txt files in cookbooks/ and benchmarks/ are now identical (except for the screen output).

tjhei commented on CMakeLists.txt:
```diff
@@ -793,14 +793,21 @@ INSTALL(DIRECTORY include/ ${CMAKE_BINARY_DIR}/include/
-  DESTINATION benchmarks
-  COMPONENT examples
-  FILES_MATCHING PATTERN "*")
+SET(ASPECT_INSTALL_EXAMPLES OFF CACHE BOOL "If ON all cookbooks and benchmarks will be built and installed.")
```
Should we call this ASPECT_ENABLE_EXAMPLES instead? (it involves building and installing)

The current name is ok as well, so I leave this up to you.

gassmoeller replied:
I think I prefer the current name. ENABLE implies to me that I cannot use the cookbooks and benchmarks if it is turned off, but the folders will always be there in the source directory. And building the examples without installing is actually not very helpful, because the libraries will be in a sub-directory of the build directory, where the .prm files will not find them.

tjhei commented:
Do you want to update documentation in a separate PR?

gassmoeller commented:
> Do you want to update documentation in a separate PR?

Yes, I can make a separate PR for that.
